tests/php/src/RemoteRequest/CachedRemoteGetRequestTest.php

```diff
@@ -0,0 +1,94 @@
+class CachedRemoteGetRequestTest extends WP_UnitTestCase {
+
+	/** @var CachedRemoteGetRequest */
+	private $instance;
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->instance = new CachedRemoteGetRequest( new StubbedRemoteGetRequest( [] ) );
+	}
+
+	/** @covers ::get() */
+	public function test_get_with_serialized_cached_response() {
+		$url         = 'https://example.com/foo.css';
+		$body        = 'body {color:red}';
+		$headers     = [ 'content-type' => 'text/css' ];
+		$status_code = 418;
+		$expiry      = new DateTimeImmutable( '+ ' . DAY_IN_SECONDS . ' seconds' );
+
+		$cache_key = CachedRemoteGetRequest::TRANSIENT_PREFIX . md5( CachedRemoteGetRequest::class . $url );
+
+		$cached_response = new CachedResponse( $body, $headers, $status_code, $expiry );
+		add_filter(
+			"pre_transient_${cache_key}",
+			static function () use ( $cached_response ) {
+				return serialize( $cached_response ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+			}
+		);
+
+		$cached_remote_request  = new CachedRemoteGetRequest( new StubbedRemoteGetRequest( [] ) );
+		$cached_remote_response = $cached_remote_request->get( $url );
+
+		$this->assertInstanceOf( RemoteGetRequestResponse::class, $cached_remote_response );
+		$this->assertSame( $body, $cached_remote_response->getBody() );
+		$this->assertSame( $headers, $cached_remote_response->getHeaders() );
+		$this->assertSame( $status_code, $cached_remote_response->getStatusCode() );
+	}
+
+	/** @covers ::get() */
+	public function test_get_with_unserialized_cached_response() {
+		$url         = 'https://example.com/foo.css';
+		$body        = 'body {color:red}';
+		$headers     = [ 'content-type' => 'text/css' ];
+		$status_code = 418;
+		$expiry      = new DateTimeImmutable( '+ ' . DAY_IN_SECONDS . ' seconds' );
+
+		$cache_key = CachedRemoteGetRequest::TRANSIENT_PREFIX . md5( CachedRemoteGetRequest::class . $url );
+
+		$cached_response = new CachedResponse( $body, $headers, $status_code, $expiry );
+		add_filter(
+			"pre_transient_${cache_key}",
+			static function () use ( $cached_response ) {
+				return $cached_response;
+			}
+		);
+
+		$cached_remote_request  = new CachedRemoteGetRequest( new StubbedRemoteGetRequest( [] ) );
+		$cached_remote_response = $cached_remote_request->get( $url );
+
+		$this->assertInstanceOf( RemoteGetRequestResponse::class, $cached_remote_response );
+		$this->assertSame( $body, $cached_remote_response->getBody() );
+		$this->assertSame( $headers, $cached_remote_response->getHeaders() );
+		$this->assertSame( $status_code, $cached_remote_response->getStatusCode() );
+	}
+
+	/** @covers ::get() */
+	public function test_get_without_cached_response() {
+		$url         = 'https://example.com/foo.css';
+		$body        = 'body {color:red}';
+		$headers     = [ 'content-type' => 'text/css' ];
+		$status_code = 200;
+
+		$remote_request         = new StubbedRemoteGetRequest( [ $url => $body ] );
```

Comment:
Something's not quite right here as I've demonstrated with failing tests in f7b1b850ad0342005894077ac5c1fc3bf06fb0f8. The `$headers` aren't being set anywhere.

Reply:
Right that's because `StubbedRemoteGetRequest` only simulates the [response body](https://github.com/ampproject/amp-toolbox-php/blob/9e7765201d9c33971879001f8a4f4a7780b49b8a/src/RemoteRequest/StubbedRemoteGetRequest.php#L42-L48). It would be most appropriate to use the `pre_http_request` filter you mentioned earlier for this test then.

Reply:
Used the `pre_http_request` filter to mock the HTTP request in https://github.com/ampproject/amp-wp/pull/6506/commits/14e46b4c4d4b4928c38385d2d186c143acf9434d.